src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Specialized;
+using System.Globalization;
+
+namespace System.Net
+{
+    internal class HeaderInfoTable
+    {
+        private static Hashtable s_headerHashTable;
+        private static HeaderInfo s_unknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, s_singleParser);
+        private static HeaderParser s_singleParser = new HeaderParser(ParseSingleValue);
+        private static HeaderParser s_multiParser = new HeaderParser(ParseMultiValue);
+
+        private static string[] ParseSingleValue(string value)
+        {
+            return new string[1] { value };
+        }
+
+        private static string[] ParseMultiValue(string value)
+        {
+            StringCollection tempStringCollection = new StringCollection();
+
+            bool inquote = false;
+            int chIndex = 0;
+            char[] vp = new char[value.Length];
+            string singleValue;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                {
+                    inquote = !inquote;
+                }
+                else if ((value[i] == ',') && !inquote)
+                {
+                    singleValue = new string(vp, 0, chIndex);
+                    tempStringCollection.Add(singleValue.Trim());
+                    chIndex = 0;
+                    continue;
+                }
+                vp[chIndex++] = value[i];
+            }
+
+            // Now add the last of the header values to the stringtable.
+
+            if (chIndex != 0)
+            {
+                singleValue = new string(vp, 0, chIndex);
+                tempStringCollection.Add(singleValue.Trim());
```

Comment:
If the finding of the starting/ending whitespace were done on ```vp``` rather than on the resulting string, we could save a string allocation here in some cases by just passing in the "right" offset/count to string's ctor and avoiding the Trim.